recordio/proto/proto_reader.go

```diff
@@ -36,27 +38,75 @@ func (r *Reader) Close() error {
 	return r.reader.Close()
 }
 
-func NewProtoReaderWithPath(path string) (ReaderI, error) {
-	f, err := os.Open(path)
-	if err != nil {
-		return nil, err
+// options
+
+type ReaderOptions struct {
+	path         string
+	file         *os.File
+	bufSizeBytes int
+}
+
+type ReaderOption func(*ReaderOptions)
+
+func ReaderPath(p string) ReaderOption {
+	return func(args *ReaderOptions) {
+		args.path = p
 	}
+}
 
-	r, err := NewProtoReaderWithFile(f)
-	if err != nil {
-		return nil, err
+func ReaderFile(p *os.File) ReaderOption {
+	return func(args *ReaderOptions) {
+		args.file = p
 	}
+}
 
-	return r, nil
+func ReadBufferSizeBytes(p int) ReaderOption {
+	return func(args *ReaderOptions) {
+		args.bufSizeBytes = p
+	}
 }
 
-func NewProtoReaderWithFile(file *os.File) (ReaderI, error) {
-	reader, err := recordio.NewFileReaderWithFile(file)
+// create a new reader with the given options. Either Path or File must be supplied
+func NewReader(readerOptions ...ReaderOption) (ReaderI, error) {
+	opts := &ReaderOptions{
+		path:         "",
+		file:         nil,
+		bufSizeBytes: 1024 * 1024 * 4,
+	}
+
+	for _, readerOption := range readerOptions {
+		readerOption(opts)
+	}
+
+	if (opts.file != nil) && (opts.path != "") {
+		return nil, errors.New("either os.File or string path must be supplied, never both")
+	}
+
+	if opts.file == nil {
+		if opts.path == "" {
+			return nil, errors.New("path was not supplied")
+		}
+	}
```

Comment:
don't forget to add some tests for the logic

Reply:
I would leave it around, it's similar to how the other readers and writes are constructed